src/trace.rs

```diff
@@ -1,120 +1,238 @@
+        Span {
+            id: tl.new_span_id(),
+            state: State::Normal,
+            parent_id: 0,
+            begin_cycles: minstant::now(),
             elapsed_cycles: 0,
-            event: pending_event.into(),
+            event: root_event.into(),
         },
-        settle_event.into(),
+        tl,
+    );
+
+    let (tx, rx) = crossbeam::channel::unbounded();
+    tl.cur_collector = Some(Arc::new(tx));
+
+    (
+        ScopeGuard { inner: span_inner },
+        Collector::new(trace_id, rx),
     )
-    .unwrap(); // It's safe to unwrap because the collector always exists at present.
+}
+
+pub fn new_span<T: Into<u32>>(event: T) -> Option<SpanGuard> {
+    let trace = TRACE_LOCAL.with(|trace| trace.get());
+    let tl = unsafe { &mut *trace };
+
+    if tl.enter_stack.is_empty() {
+        return None;
+    }
 
-    (trace_guard, collector)
+    let parent_id = *tl.enter_stack.last().unwrap();
+    let span_inner = SpanGuardInner::enter(
+        Span {
+            id: tl.new_span_id(),
+            state: State::Normal,
+            parent_id,
+            begin_cycles: minstant::now(),
+            elapsed_cycles: 0,
+            event: event.into(),
+        },
+        tl,
+    );
+
+    Some(SpanGuard { inner: span_inner })
 }
 
-#[must_use]
-#[inline]
-pub fn trace_may_enable<T: Into<u32>>(
-    enable: bool,
-    event: T,
-) -> (
-    Option<crate::trace_local::LocalTraceGuard>,
-    Option<crate::collector::Collector>,
-) {
-    let event = event.into();
-    trace_may_enable_fine(enable, event, event)
+/// The property is in bytes format, so it is not limited to be a key-value pair but
+/// anything intended. However, the downside of flexibility is that manual encoding
+/// and manual decoding need to consider.
+pub fn new_property<B: AsRef<[u8]>>(p: B) {
+    append_property(|| p);
 }
 
-#[must_use]
-#[inline]
-pub fn trace_may_enable_fine<E1: Into<u32>, E2: Into<u32>>(
-    enable: bool,
-    pending_event: E1,
-    settle_event: E2,
-) -> (
-    Option<crate::trace_local::LocalTraceGuard>,
-    Option<crate::collector::Collector>,
-) {
-    if enable {
-        let (guard, collector) = trace_enable_fine(pending_event, settle_event);
-        (Some(guard), Some(collector))
-    } else {
-        (None, None)
+/// `property` of closure version
+pub fn new_property_with<F, B>(f: F)
+where
+    B: AsRef<[u8]>,
+    F: FnOnce() -> B,
+{
+    append_property(f);
+}
+
+pub struct TraceLocal {
+    /// For id construction
+    pub id_prefix: u16,
+    pub id_suffix: u16,
+
+    /// For parent-child relation construction. The last span, when exits, is
+    /// responsible to submit the local span sets.
+    pub enter_stack: Vec<SpanId>,
+    pub span_set: SpanSet,
+
+    pub cur_collector: Option<Arc<Sender<SpanSet>>>,
+}
+
+impl TraceLocal {
+    #[inline]
+    pub fn new_span_id(&mut self) -> SpanId {
+        let id = ((self.id_prefix as u32) << 16) | self.id_suffix as u32;
```

Comment:
I think we need to prevent `new_span_id` from generating id `0` because a `0` placed at `parent_id` indicates the span is a root which is not doubt a ruin.

Reply:
span_id will not be 0:

https://github.com/pingcap-incubator/minitrace-rust/blob/2e41c00f6fde91f57300fcba1aca77d62d10d030/src/trace.rs#L13

Reply:
`fetch_add` is wrapping add

Reply:
> `fetch_add` is wrapping add

you're unlikely to run out of u32 tho... if that could be the case, there would be lots more id collision besides of this one.

Reply:
just let `id_suffix` never be zero

L117 -> `self.id_suffix = 1;`
L18 -> `id_suffix = 1`